pkg/addonmanager/cloudevents/manager.go

```diff
@@ -0,0 +1,289 @@
+package cloudevents
+
+import (
+	"context"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addonv1alpha1client "open-cluster-management.io/api/client/addon/clientset/versioned"
+	addoninformers "open-cluster-management.io/api/client/addon/informers/externalversions"
+	clusterv1client "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	clusterv1informers "open-cluster-management.io/api/client/cluster/informers/externalversions"
+	workinformers "open-cluster-management.io/api/client/work/informers/externalversions"
+
+	"open-cluster-management.io/addon-framework/pkg/addonmanager"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/addonconfig"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/agentdeploy"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/certificate"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/cmaconfig"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/cmamanagedby"
+	"open-cluster-management.io/addon-framework/pkg/addonmanager/controllers/registration"
+	"open-cluster-management.io/addon-framework/pkg/basecontroller/factory"
+	"open-cluster-management.io/addon-framework/pkg/index"
+	"open-cluster-management.io/addon-framework/pkg/utils"
+	cloudeventswork "open-cluster-management.io/sdk-go/pkg/cloudevents/work"
+	"open-cluster-management.io/sdk-go/pkg/cloudevents/work/source/codec"
+)
+
+// cloudeventsAddonManager is the implementation of AddonManager with
+// the base implementation and cloudevents options
+type cloudeventsAddonManager struct {
```

Comment:
could you explain the difference between the cloudevents addon manager and the classic addon manager except for the building way of manifestwork client and informer? which I think is good for review.

Reply:
the difference between the cloudevents addon manager and the classic addon manager is the way to build work client and work informer.

the cloudevents addon manager use cloudevents drivers(mqtt or grpc) to create work informer, so that the manifestwork used to deploy addon agent can be delivered to managed cluster through mqtt or grpc.

no other difference.